### ISSUE TITLE
Do C callbacks require extern "C"?

### ISSUE THREAD

**Issue by ivan-pi** [fortran-lang/minpack#78]
Assuming the [C interface](https://github.com/fortran-lang/minpack/blob/main/include/minpack.h) of MINPACK is used in C++ program, should the callback types also have C linkage specified?
```cpp
extern "C" {                  // <-- do we need this?
typedef void (*minpack_func)(
    int /* n */,
    const double* /* x */,
    double* /* fvec */,
    int* /* iflag */,
    void* /* udata */);
}                             // <-- do we need this?
```

The last section of this [Oracle document](https://docs.oracle.com/cd/E19059-01/wrkshp50/805-4956/bajdcjch/index.html), _Pointers to Functions as Function Parameters_, appears to say that unless the linkage is specified, the following declaration
```cpp
MINPACK_EXTERN void MINPACK_CALL
minpack_hybrd(
    minpack_func /* fcn */,
    int /* n */,
    double* /* x */,
    double* /* fvec */,
    ...);
```
assumes the `fcn` argument will have C++ linkage and not C, like expected by the Fortran interfaces (definitions). 

Typically, C and C++ compilers of the same vendor will be binary compatible when it comes to function linkage, but in principle they don't have to be.

**Comment by certik** [comment]
Yes, you have to put in `extern "C"` for C++. You typically put it under defines to only be used in C++ codes, not C.

**Comment by ivan-pi** [comment]
We have that in `MINPACK_EXTERN` macro for the procedures, but not for the callback typedefs. According to the Oracle resources, the typedefs should have it too.

**Comment by certik** [comment]
Ah I see. Looks good to me.
